third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -2813,112 +2825,281 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
-                        .map(|(_, value)| value)
-                        .collect_vec();
-                    Some((struct_id, args))
-                }
+    ) -> Option<()> {
+        let mut succeed = true;
+        let mut fields_not_covered: BTreeSet<Symbol> = BTreeSet::new();
+        // Exclude from the covered fields the dummy_field added by legacy compiler
+        fields_not_covered.extend(
+            field_decls
+                .keys()
+                .filter(|s| *s != &self.parent.dummy_field_name()),
+        );
+        for (name_loc, name, (_, _)) in fields.iter() {
+            let field_name = self.symbol_pool().make(name);
+            if let Some((_, _, _)) = field_decls.get(&field_name) {
+                fields_not_covered.remove(&field_name);
             } else {
                 self.error(
-                    &struct_name_loc,
+                    &self.to_loc(&name_loc),
                     &format!(
-                        "native struct `{}` cannot be packed or unpacked",
+                        "field `{}` not declared in struct `{}`",
+                        field_name.display(self.symbol_pool()),
                         struct_name.display(self.parent.parent.env)
                     ),
                 );
-                None
+                succeed = false;
             }
-        } else {
+        }
+        if !fields_not_covered.is_empty() {
             self.error(
-                &struct_name_loc,
+                loc,
                 &format!(
-                    "undeclared struct `{}`",
-                    struct_name.display(self.parent.parent.env)
+                    "missing fields {}",
+                    fields_not_covered
+                        .iter()
+                        .map(|n| format!("`{}`", n.display(self.symbol_pool())))
+                        .join(", ")
                 ),
             );
+            succeed = false;
+        }
+        if succeed {
+            Some(())
+        } else {
             None
         }
     }
 
+    // return the indices of fields that can be left in place during transforming pack exprs
+    fn in_order_fields<T>(
+        &mut self,
+        field_decls: &BTreeMap<Symbol, (Loc, usize, Type)>,
+        fields: &EA::Fields<T>,
+    ) -> BTreeSet<usize> {
+        // maps def_idx to exp_idx
+        let mut permutation = BTreeMap::new();
+        for (_, name, (exp_idx, _)) in fields.iter() {
+            let field_name = self.symbol_pool().make(name);
+            let (_, def_idx, _) = field_decls.get(&field_name).unwrap();
+            permutation.insert(*def_idx, *exp_idx);
+        }
+        let permutation = permutation
+            .into_iter()
+            .sorted_by_key(|(i, _)| *i)
+            .map(|(_, value)| value)
+            .collect_vec();
+        let mut in_order_fields = BTreeSet::new();
+        if !permutation.is_empty() {
+            let mut prev = permutation[0];
+            in_order_fields.insert(0);
```

Comment:
Some comment explaining how this works or citing some standard algorithm would be useful...